### PR TITLE
feat: snapshot apis for EIP-4881

### LINKED ISSUE #293: Snapshot apis
**Is your feature request related to a problem? Please describe.**

This is to support EIP-4881 https://eips.ethereum.org/EIPS/eip-4881#deposit-finalization-conditions

1. `get_snapshot`: API to extract finalized roots given a gindex (or depth + index) (below is for depth = 4 and deposit_count = index)

Screen Shot 2022-12-23 at 15 33 55

2. `from_snapshot`: API to reconstruct the tree from a snapshot (similar to reconstructing from a proof), params would be a list of finalized roots and a gindex (or depth + index)

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

Implement snapshot apis for eip-4881

**Description**

- persistent-merkle-tree: implement toSnapshot/fromSnapshot apis
- ssz:
  - define new `PartialListCompositeType` type along with ViewDU (and no View since we lacks a lot of methods there and dont maintain it anymore)
  - new `toPartialViewDU()` api as the entry point
  - consume persistent-merkle-tree `toSnapshot, fromSnapshot` apis
 - notes for consumers (lodestar)
   - new ` DepositDataRootPartialList = new PartialListCompositeType(Root, 2 ** DEPOSIT_CONTRACT_TREE_DEPTH);`
   - download snapshot
   - then create the deposit root tree like `depositRootTree = DepositDataRootPartialList.toPartialViewDU(snapshot)` 

Closes #293
a candidate to replace #377

**TODOs**
- [x] spec tests

### CODE REVIEW

File: packages/ssz/src/viewDU/partialListComposite.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+  // push() is inherited from ListCompositeTreeViewDU
+
+  /**
+   * Similar to ListCompositeTreeViewDU.sliceTo() but:
+   * - throw error if index < snapshot.count - 1
+   * - special handle for index === snapshot.count - 1, we restore from snapshot
+   */
+  sliceTo(index: number): this {
+    if (index < this.snapshot.count - 1) {
```
Comment (g11tech) [inline packages/ssz/src/viewDU/partialListComposite.ts:89]: nice

**Review by g11tech — DISMISSED** [review]
looks super amazing :+1: 

i think we will now be saving the `DepositTree` in db by the finalized gindex nodes and then values post that?

**Comment by twoeths** [comment]
> looks super amazing 👍
> 
> i think we will now be saving the `DepositTree` in db by the finalized gindex nodes and then values post that?

now instead of saving all roots, we'll just save serialized snapshot periodically. Something like:

```typescript
const treeSnapshot = depositRootTree.toSnapshot(depositCount);
const depositSnapshot = {
    finalized: treeSnapshot.finalized,
    depositRoot: treeSnapshot.root,
    depositCount: treeSnapshot.count,
    executionBlockHash,
    executionBlockHeight,
};
db.put(executionBlockHeight, depositSnapshot);
```
upon restart we download snapshot from `checkpointSyncUrl`, compare to db's snapshot and use either of them, then reconstruct the partial tree from snapshot
```typescript
const depositRootTree = DepositDataRootPartialList.toPartialViewDU({
  finalized: snapshot.finalized,
  root: snapshot.depositRoot,
  count: snapshot.depositCount,
});
```

then the partial tree grows from there similar to the current full tree

File: packages/ssz/src/util/types.ts [hunk 2]
```diff
@@ -1 +1,8 @@
 export type Require<T, K extends keyof T> = T & Required<Pick<T, K>>;
+
+/** Snapshot for PartialListCompositeType */
```
Comment (wemeetagain) [inline packages/ssz/src/util/types.ts:3]: Can you add more comments here?

Eg:
A snapshot contains the minimum amount of information needed to reconstruct a merkleized list, for the purposes of appending more items. Note: This does not contain list elements, rather only contains intermediate merkle nodes.

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**